File: backend/app/schemas/exports.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class CompanySqlExportRead(BaseModel):
    export_job_id: str
    daily_report_id: str
    workspace_code: str
    domain_code: str
    status: str
    item_count: int
    statement_count: int
    sql_text: str
    sql_text_bytes: int
    sql_text_preview_bytes: int
    sql_text_truncated: bool
    download_url: str
    download_filename: str
    created_at: datetime
    completed_at: datetime | None
    result_json: dict[str, Any]


class CompanySqlBatchExportCreate(BaseModel):
    daily_report_ids: list[str] = Field(min_length=1, max_length=31)
    continue_on_error: bool = True


class CompanySqlBatchExportItemRead(BaseModel):
    daily_report_id: str
    day_key: str | None
    status: str
    preflight_status: str
    export_job_id: str | None
    download_url: str | None
    item_count: int
    statement_count: int
    sql_text_bytes: int
    warning_count: int
    error_count: int
    errors: list[str]


class CompanySqlBatchExportRead(BaseModel):
    batch_export_job_id: str
    workspace_code: str
    domain_code: str
    status: str
    total_reports: int
    succeeded_count: int
    failed_count: int
    skipped_count: int
    total_item_count: int
    total_statement_count: int
    total_sql_text_bytes: int
    manifest_json: dict[str, Any]
    items: list[CompanySqlBatchExportItemRead]
    created_at: datetime
    completed_at: datetime | None


class CompanySqlPreflightIssueRead(BaseModel):
    level: str
    code: str
    message: str
    field: str | None = None
    daily_report_item_id: str | None = None


class CompanySqlPreflightItemRead(BaseModel):
    daily_report_item_id: str
    generated_news_id: str
    news_item_id: str
    adoption_status: int
    status: str
    title: str
    source_url: str | None
    category: str | None
    errors: list[CompanySqlPreflightIssueRead]
    warnings: list[CompanySqlPreflightIssueRead]


class CompanySqlPreflightRead(BaseModel):
    daily_report_id: str
    workspace_code: str
    domain_code: str
    day_key: str
    report_status: str
    status: str
    eligible_count: int
    blocked_count: int
    skipped_count: int
    warning_count: int
    error_count: int
    errors: list[CompanySqlPreflightIssueRead]
    warnings: list[CompanySqlPreflightIssueRead]
    items: list[CompanySqlPreflightItemRead]


CompanySqlImportStatus = Literal["pending", "imported", "failed", "partial"]


class CompanySqlImportFailureCreate(BaseModel):
    export_job_item_id: str | None = None
    sql_sequence: int | None = Field(default=None, ge=1)
    sql_table: str | None = None
    error_code: str = ""
    error_message: str = Field(min_length=1, max_length=2000)
    sql_excerpt: str = ""


class CompanySqlImportReceiptCreate(BaseModel):
    target_system: str = Field(default="company_intranet", min_length=1, max_length=128)
    import_status: CompanySqlImportStatus
    imported_at: datetime | None = None
    imported_statement_count: int = Field(default=0, ge=0)
    failed_statement_count: int = Field(default=0, ge=0)
    failure_items: list[CompanySqlImportFailureCreate] = Field(default_factory=list, max_length=200)
    notes: str = Field(default="", max_length=4000)


class CompanySqlImportFailureRead(BaseModel):
    export_job_item_id: str | None
    sql_sequence: int | None
    sql_table: str | None
    error_code: str
    error_message: str
    sql_excerpt: str


class CompanySqlImportReceiptRead(BaseModel):
    id: str
    export_job_id: str
    workspace_code: str
    domain_code: str
    target_system: str
    import_status: str
    imported_at: datetime | None
    imported_statement_count: int
    failed_statement_count: int
    failure_items: list[CompanySqlImportFailureRead]
    notes: str
    recorded_by_id: str | None
    recorded_by_name: str | None
    created_at: datetime
    updated_at: datetime


class ExportJobRead(BaseModel):
    id: str
    export_type: str
    status: str
    workspace_code: str
    domain_code: str
    params_json: dict[str, Any]
    result_json: dict[str, Any]
    latest_import_receipt: CompanySqlImportReceiptRead | None = None
    created_at: datetime
    completed_at: datetime | None


class CompanySqlTraceFieldDiffRead(BaseModel):
    field: str
    label: str
    export_source: str
    export_value_preview: str
    generated_value_preview: str | None
    editor_value_preview: str | None
    raw_value_preview: str | None
    changed_by_editor: bool
    truncated: bool


class CompanySqlTraceItemRead(BaseModel):
    export_job_item_id: str
    sql_sequence: int
    sql_table: str
    status: str
    daily_report_item_id: str
    generated_news_id: str
    news_item_id: str
    raw_item_id: str | None
    data_source_id: str | None
    data_source_name: str | None
    source_type: str | None
    source_url: str | None
    source_title: str
    generated_title: str
    export_title: str
    category: str
    adoption_status: int
    sql_excerpt: str
    title_source: str
    summary_source: str
    key_points_source: str
    content_field_sources: dict[str, str]
    editor_override_fields: list[str]
    field_diffs: list[CompanySqlTraceFieldDiffRead]


class CompanySqlTraceRead(BaseModel):
    export_job_id: str
    item_count: int
    statement_count: int
    trace_items: list[CompanySqlTraceItemRead]

File: frontend/src/api/exports.ts
import { requestBlob, requestJson } from "./http";

export interface CompanySqlExportRecord {
  export_job_id: string;
  daily_report_id: string;
  workspace_code: string;
  domain_code: string;
  status: string;
  item_count: number;
  statement_count: number;
  sql_text: string;
  sql_text_bytes: number;
  sql_text_preview_bytes: number;
  sql_text_truncated: boolean;
  download_url: string;
  download_filename: string;
  created_at: string;
  completed_at: string | null;
  result_json: Record<string, unknown>;
}

export interface CompanySqlBatchExportItemRecord {
  daily_report_id: string;
  day_key: string | null;
  status: string;
  preflight_status: string;
  export_job_id: string | null;
  download_url: string | null;
  item_count: number;
  statement_count: number;
  sql_text_bytes: number;
  warning_count: number;
  error_count: number;
  errors: string[];
}

export interface CompanySqlBatchExportRecord {
  batch_export_job_id: string;
  workspace_code: string;
  domain_code: string;
  status: string;
  total_reports: number;
  succeeded_count: number;
  failed_count: number;
  skipped_count: number;
  total_item_count: number;
  total_statement_count: number;
  total_sql_text_bytes: number;
  manifest_json: Record<string, unknown>;
  items: CompanySqlBatchExportItemRecord[];
  created_at: string;
  completed_at: string | null;
}

export interface ExportJobRecord {
  id: string;
  export_type: string;
  status: string;
  workspace_code: string;
  domain_code: string;
  params_json: Record<string, unknown>;
  result_json: Record<string, unknown>;
  latest_import_receipt: CompanySqlImportReceiptRecord | null;
  created_at: string;
  completed_at: string | null;
}

export interface CompanySqlImportFailureRecord {
  export_job_item_id: string | null;
  sql_sequence: number | null;
  sql_table: string | null;
  error_code: string;
  error_message: string;
  sql_excerpt: string;
}

export interface CompanySqlImportReceiptRecord {
  id: string;
  export_job_id: string;
  workspace_code: string;
  domain_code: string;
  target_system: string;
  import_status: string;
  imported_at: string | null;
  imported_statement_count: number;
  failed_statement_count: number;
  failure_items: CompanySqlImportFailureRecord[];
  notes: string;
  recorded_by_id: string | null;
  recorded_by_name: string | null;
  created_at: string;
  updated_at: string;
}

export interface CompanySqlImportReceiptCreatePayload {
  target_system: string;
  import_status: "pending" | "imported" | "failed" | "partial";
  imported_at?: string | null;
  imported_statement_count: number;
  failed_statement_count: number;
  failure_items: Array<{
    export_job_item_id?: string | null;
    sql_sequence?: number | null;
    sql_table?: string | null;
    error_code?: string;
    error_message: string;
    sql_excerpt?: string;
  }>;
  notes: string;
}

export interface CompanySqlPreflightIssueRecord {
  level: string;
  code: string;
  message: string;
  field: string | null;
  daily_report_item_id: string | null;
}

export interface CompanySqlPreflightItemRecord {
  daily_report_item_id: string;
  generated_news_id: string;
  news_item_id: string;
  adoption_status: number;
  status: string;
  title: string;
  source_url: string | null;
  category: string | null;
  errors: CompanySqlPreflightIssueRecord[];
  warnings: CompanySqlPreflightIssueRecord[];
}

export interface CompanySqlPreflightRecord {
  daily_report_id: string;
  workspace_code: string;
  domain_code: string;
  day_key: string;
  report_status: string;
  status: string;
  eligible_count: number;
  blocked_count: number;
  skipped_count: number;
  warning_count: number;
  error_count: number;
  errors: CompanySqlPreflightIssueRecord[];
  warnings: CompanySqlPreflightIssueRecord[];
  items: CompanySqlPreflightItemRecord[];
}

export interface CompanySqlTraceItemRecord {
  export_job_item_id: string;
  sql_sequence: number;
  sql_table: string;
  status: string;
  daily_report_item_id: string;
  generated_news_id: string;
  news_item_id: string;
  raw_item_id: string | null;
  data_source_id: string | null;
  data_source_name: string | null;
  source_type: string | null;
  source_url: string | null;
  source_title: string;
  generated_title: string;
  export_title: string;
  category: string;
  adoption_status: number;
  sql_excerpt: string;
  title_source: string;
  summary_source: string;
  key_points_source: string;
  content_field_sources: Record<string, string>;
  editor_override_fields: string[];
  field_diffs: CompanySqlTraceFieldDiffRecord[];
}

export interface CompanySqlTraceFieldDiffRecord {
  field: string;
  label: string;
  export_source: string;
  export_value_preview: string;
  generated_value_preview: string | null;
  editor_value_preview: string | null;
  raw_value_preview: string | null;
  changed_by_editor: boolean;
  truncated: boolean;
}

export interface CompanySqlTraceRecord {
  export_job_id: string;
  item_count: number;
  statement_count: number;
  trace_items: CompanySqlTraceItemRecord[];
}

export async function fetchExportJobs(workspaceCode?: string): Promise<ExportJobRecord[]> {
  const params = workspaceCode ? `?${new URLSearchParams({ workspace_code: workspaceCode }).toString()}` : "";
  return requestJson<ExportJobRecord[]>(`/api/exports${params}`);
}

export async function fetchExportJob(exportJobId: string): Promise<ExportJobRecord> {
  return requestJson<ExportJobRecord>(`/api/exports/${exportJobId}`);
}

export async function fetchCompanySqlTrace(exportJobId: string): Promise<CompanySqlTraceRecord> {
  return requestJson<CompanySqlTraceRecord>(`/api/exports/${exportJobId}/trace`);
}

export async function fetchCompanySqlImportReceipts(exportJobId: string): Promise<CompanySqlImportReceiptRecord[]> {
  return requestJson<CompanySqlImportReceiptRecord[]>(`/api/exports/${exportJobId}/import-receipts`);
}

export async function createCompanySqlImportReceipt(
  exportJobId: string,
  payload: CompanySqlImportReceiptCreatePayload
): Promise<CompanySqlImportReceiptRecord> {
  return requestJson<CompanySqlImportReceiptRecord>(`/api/exports/${exportJobId}/import-receipts`, {
    method: "POST",
    body: JSON.stringify(payload)
  });
}

export async function downloadExportJob(exportJobId: string): Promise<Blob> {
  return requestBlob(`/api/exports/${exportJobId}/download`);
}

export async function preflightCompanySqlExport(dailyReportId: string): Promise<CompanySqlPreflightRecord> {
  return requestJson<CompanySqlPreflightRecord>(`/api/exports/company-sql/daily-reports/${dailyReportId}/preflight`, {
    method: "POST"
  });
}

export async function createCompanySqlExport(dailyReportId: string): Promise<CompanySqlExportRecord> {
  return requestJson<CompanySqlExportRecord>(`/api/exports/company-sql/daily-reports/${dailyReportId}`, {
    method: "POST"
  });
}

export async function createCompanySqlBatchExport(dailyReportIds: string[]): Promise<CompanySqlBatchExportRecord> {
  return requestJson<CompanySqlBatchExportRecord>("/api/exports/company-sql/daily-reports/batch", {
    method: "POST",
    body: JSON.stringify({ daily_report_ids: dailyReportIds, continue_on_error: true })
  });
}
